Refaktor

kytomaki opened the pull request:
This will

- [x] Add structs to handle certificates for easier testing
- [x] Wrap the whole thing in [cobra](https://github.com/spf13/cobra)
- [x] Rebase and clean up for easier future updates
- [x] Update readme
- [x] Rework packaging

Stuff that has been left for later iterations

- [ ] ~~Generate man page~~
- [ ] ~~Generate bash completion~~

On merge this should tagged `2.0.0` since the usage has changed.

Commit:
Switch from glide to dep

Commit:
Clean up Makefile

Commit:
Refactor keepsake to use cobra

* As in the spf13/cobra
* Remove openssl from constraints, we got rid of spacmonkeygo's openssl
  implementation
* Use vault 1.0.0

zircote commented:
@kytomaki I added @etotten and @ilapost to the review.

etotten commented:
I'll dig into this Friday morning (sorry I wasn't able to today)

etotten commented:
Actually, from talking to @kytomaki, it sounds like this might still be in somewhat early stages, so holding off on a detailed review.

Commit:
Moved cmd/testdata to testdata.

Removed all crt csr and key files from repo and added them in .gitignore.
Minor refactors to config.go and update.go.

* Write bundle cert to correct location

Commit:
Finished update implementation.

Initial commit of this resulted in problems in handling the net.IP.
This now contains a commit which changed the `IPAddresses` field of the
config struct into a slice of strings.

* Use pointers to certificateconfs

Very small but quite influential change. Without the pointers there is a
great risk of clashing configurations. Which usually is a cause for
great anguish and gnashing of teeth.

* Refactor to use own error type to distinguish recoverable situations

* Handle vault responses as strings

Also fix shadowed variable

* Handle non existing files in the beginning

* Refactor validity testing logic in update

* Add logging to vault

* Refactor inner variable naming for better readability

* Handle external command on update

* Add thread safe way of getting singleton vault client

* Extract read and update into methods for later use

* Add mutex to certificate retrieval

This should help prevent possible race conditions while fetching
certificates from vault.

* Prune away some dead code

* Handle errors to satisfy go-metalinter

kytomaki committed:
Add configurable log level and type

Currently type only supports type `json`

kytomaki committed:
Add command to print out version info

Also refactor build time variables out of main package.

kytomaki committed:
Add daemon sub-command for update

This should allow us to run the updates indefinitely.

Add signal handling to daemon command

The program will:
* exit on KILL and TERM
* reload on HUP

Write certificate files on exiting the loop

Add return channel for daemon

Simplify writefiles function

* Add helper functions to write specific files
* Rework encoding of certs

* Add function to determine if we have a certificate bundle

This should make the reading a bit easier.

* Handle stopping of watching in a function

This also fixes the signal handling.

kytomaki committed:
Update documentation

* Add some informational echos to Makefile
* Add an example of configuration
* Add more documentation on running the tests in container

etotten commented on cmd/update.go:
```diff
@@ -0,0 +1,63 @@
+var updateCmd = &cobra.Command{
+	Use:   "update",
+	Short: "Update certificates from vault",
+	Long:  `Check for validity of certificates and update if needed`,
```
Should be about ``update``, not ``check``

etotten commented on cmd/config.go:
```diff
@@ -0,0 +1,240 @@
+			certDuration := cert.NotAfter.Sub(cert.NotBefore)
+			tresholdTime := cert.NotBefore.Add(MultipliedDuration(certDuration, Conf.RenewalCoefficient))
+			if time.Now().After(tresholdTime) {
+				return RecoverableKeepsakeError(fmt.Sprintf("certificate expiration time %s, is over treshold %s", cert.NotAfter, tresholdTime))
```
Misspelling: "threshold" (here and in the variable name)

etotten commented on Makefile:
```diff
@@ -148,38 +56,23 @@ docker-clean:
+	bin/keepsake update 
 
 .PHONY: docker-integration-test
 docker-integration-test: docker/Dockerfile_amzn docker/Dockerfile_vault
```
Maybe a comment here (or wherever most relevant) stating what one can expect of this test. I ran it locally, and saw it execute stuff, but didn't know if it passed or not. So, is it considering passing of there are no errors output? I'm guessing that you can leave the files and run it later and expect it to update more than if you run it again quickly?

kytomaki replied:
I added some informational `echo`'s to the `docker-integration-test` target. And also updated the README

etotten commented on cmd/config.go:
```diff
@@ -0,0 +1,240 @@
+// ReadFiles read the certificates from defined files
+func (cconf *CertificateConf) ReadFiles() (err error) {
+	// Bundle file is not set
+	if cconf.BundleCertificateFile == "" {
```
Minor thing, but it might be more readable / clean if there was a isBundledCert() function that does this boolean check.

kytomaki replied:
This has also been done.

etotten commented on cmd/config.go:
```diff
@@ -0,0 +1,240 @@
+			case "ttl":
+				tests = append(tests, CheckTTL())
+			default:
+				// TODO: Unknown test key, should we return error?
```
Yes, I say we should return an error. If they ask for a test that doesn't exist, then something is wrong and will cause sadness.

kytomaki replied:
Now it returns error

etotten commented on cmd/daemon.go:
```diff
@@ -0,0 +1,119 @@
+		cconf.updateInvalidCertificate()
+		defer func() {
+			log.Infof("writing files for certificate: %s", cconf.CName)
+			if err := cconf.WriteFiles(); err != nil {
```
So, I'm probably missing something, but I see that ``UpdateInvalidCertificate()`` in update.go already calls ``WriteFiles()`` (currently line 59), so why are we doing it again?

etotten replied:
Also why is the call ``cconf.updateInvalidCertificate()`` outside of the defer but ``cconf.WriteFiles()`` is inside?

kytomaki replied:
I was getting paranoid about the concurrency. But I think you're right. I'll ditch this.

etotten reviewed (changes requested):
Went through the whole thing, made some pretty small comments, asked a couple of questions. I need to probably walk through a little bit of the logic again in the morning, but generally, I think this is looking pretty sweet.

etotten commented:
This refaktor is looking generally good. I expect it's probably time to squash, eh? I think 72 commits is a new record.

One question I have is about whether it makes sense to just use -once in an hourly (or so) cron instead of as a daemon. It should be able to simply read the certificate files and see there's nothing to do the vast majority of time, right? Would that be less likely to have issues than a bunch of long running go routines? Seems a bit simpler, though I suppose, less cool. Dunno...wanted to get your thoughts on it.

Even though I'm on vacation for a bit, I'll watch for replies / updates since this presumably needs to keep moving.

kytomaki committed:
Add test to check if TTL is valid

kytomaki committed:
Reword package target for better readability

kytomaki committed:
Use correct paths in makefile

kytomaki committed:
Use prerequisites for handling what to package

kytomaki committed:
Package configuration files with rpm

etotten reviewed (approved):
LGTM
